pkg/scheduler/queue/queue.go

```diff
@@ -464,69 +526,129 @@ func (q *RequestQueue) GetConnectedQuerierWorkersMetric() float64 {
+// are submitted from the querier to an endpoint, separate from any specific querier-worker connection.
+func (q *RequestQueue) SubmitNotifyQuerierShutdown(ctx context.Context, querierID QuerierID) {
+	// Create a generic querier-worker connection to submit the operation.
+	conn := NewUnregisteredQuerierWorkerConn(ctx, querierID) // querierID matters but workerID does not
+	q.submitQuerierWorkerOperation(conn, notifyShutdown)
+}
 
+func (q *RequestQueue) submitQuerierWorkerOperation(
+	conn *QuerierWorkerConn, opType querierOperationType,
+) {
+	op := newQuerierWorkerOperation(conn, opType)
+
+	// we do not check for a context cancel here;
+	// if the caller's context is canceled, we still want the dispatcherLoop to process the operation
+	// and update its tracking of querier-worker connections and querier shutting down / forget-delay states
 	select {
-	case q.querierOperations <- op:
+	case q.querierWorkerOperations <- op:
 		// The dispatcher has received the operation. There's nothing more to do.
 	case <-q.stopCompleted:
 		// The dispatcher stopped before it could process the operation. There's nothing more to do.
 	}
 }
 
-func (q *RequestQueue) processQuerierOperation(querierOp querierOperation) (resharded bool) {
-	switch querierOp.operation {
+func (q *RequestQueue) processQuerierWorkerOperation(querierWorkerOp *querierWorkerOperation) (resharded bool) {
+	switch querierWorkerOp.operation {
 	case registerConnection:
-		return q.processRegisterQuerierConnection(querierOp.querierID)
+		resharded = q.processRegisterQuerierWorkerConn(querierWorkerOp.conn)
 	case unregisterConnection:
-		return q.processUnregisterQuerierConnection(querierOp.querierID)
+		resharded = q.processUnregisterQuerierWorkerConn(querierWorkerOp.conn)
 	case notifyShutdown:
 		// No cleanup needed here in response to a graceful shutdown; just set querier state to shutting down.
-		// All subsequent waitingQuerierConns for the querier will receive an ErrShuttingDown.
+		// All subsequent waitingQuerierConns for the querier will receive an ErrQuerierShuttingDown.
 		// The querier-worker's end of the QuerierLoop will exit once it has received enough errors,
 		// and the Querier connection counts will be decremented as the workers disconnect.
-		return q.queueBroker.notifyQuerierShutdown(querierOp.querierID)
+		resharded = q.queueBroker.notifyQuerierShutdown(querierWorkerOp.conn.QuerierID)
 	case forgetDisconnected:
-		return q.processForgetDisconnectedQueriers()
+		resharded = q.processForgetDisconnectedQueriers()
 	default:
-		panic(fmt.Sprintf("received unknown querier event %v for querier ID %v", querierOp.operation, querierOp.querierID))
+		msg := fmt.Sprintf(
+			"received unknown querier-worker event %v for querier ID %v",
+			querierWorkerOp.operation, querierWorkerOp.conn.QuerierID,
+		)
+		panic(msg)
 	}
+	if querierWorkerOp.IsAwaitable() {
+		select {
```

Comment:
regarding "default" - the `case <-querierWorkerOp.conn.ctx.Done():` is the "default" here because it captures context cancellation from timeout on requests.

In the general case,  I cannot think of any action that's reasonable to take for a default branch when the desired result of the select is to complete the branch which is writing to an unbuffered channel - other than like logging at light speed.

This setup an exact copy of the logic in the outer select branch of `WaitForRequestForQuerier`, which  is why I used it as it's proven in a much hotter part of the loop (every dequeue) where as this only happens on querier-worker connect.

Conceivably, you could initialize the channels to be length 1 in this pattern and then the RequestQueue could move on after writing the dequeued item or querier connection update. Not the worst idea - I can't think of a reason right now why it _has_ to be synchronized - but since it's proven to work without this, I would not want to introduce that new pattern right now because we would want a bunch more benchmarking, etc.

Reply:
I'm not arguing for a default case, just a plain-English description of what's going on to convince the reader that a deadlock won't happen, and to forewarn anyone who may be implementing other awaitable operations in the future. Maybe I'm coming from too cautious a perspective, though.

Reply:
OK will add to all the places this pattern is used in the RequestQueue

Reply:
added this to all the relevant spots:

```
// context done case serves as a default case to bail out
// if the waiting querier-worker connection's context times out or is canceled,
// allowing the dispatcherLoop to proceed with its next iteration
```